File: library/src/main/java/me/proxer/library/api/user/UserMediaListEndpoint.java
package me.proxer.library.api.user;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.PagingLimitEndpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entitiy.user.UserMediaListEntry;
import me.proxer.library.enums.Category;
import me.proxer.library.enums.UserMediaListFilterType;
import me.proxer.library.enums.UserMediaListSortCriteria;

import javax.annotation.Nullable;
import java.util.List;

/**
 * Endpoint for requesting the media list of an user.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class UserMediaListEndpoint implements PagingLimitEndpoint<List<UserMediaListEntry>> {

    private final InternalApi internalApi;

    @Nullable
    private final String userId;

    @Nullable
    private final String username;

    /**
     * Sets the category to filter by.
     */
    @Nullable
    @Setter
    private Category category;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer page;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter
    private Integer limit;

    /**
     * Sets the query to search for.
     */
    @Nullable
    @Setter
    private String search;

    /**
     * Sets the query to search for only from the start.
     */
    @Nullable
    @Setter
    private String searchStart;

    /**
     * Sets the filter.
     */
    @Nullable
    @Setter
    private UserMediaListFilterType filter;

    /**
     * Sets if hentai should be included in the result.
     */
    @Nullable
    @Setter
    private Boolean includeHentai;

    /**
     * Set the criteria for sorting.
     */
    @Nullable
    @Setter
    private UserMediaListSortCriteria sort;

    UserMediaListEndpoint(final InternalApi internalApi, @Nullable final String userId,
                          @Nullable final String username) {
        if (userId == null && username == null) {
            throw new IllegalArgumentException("You must pass either an userId or an username.");
        }

        this.internalApi = internalApi;
        this.userId = userId;
        this.username = username;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ProxerCall<List<UserMediaListEntry>> build() {
        return internalApi.userMediaList(userId, username, category, page, limit, search, searchStart, filter, sort,
                includeHentai);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/user/UserMediaListEndpoint.kt
package me.proxer.library.api.user

import me.proxer.library.ProxerCall
import me.proxer.library.api.PagingLimitEndpoint
import me.proxer.library.entity.user.UserMediaListEntry
import me.proxer.library.enums.Category
import me.proxer.library.enums.UserMediaListFilterType
import me.proxer.library.enums.UserMediaListSortCriteria

/**
 * Endpoint for requesting the media list of an user.
 *
 * @author Ruben Gees
 */
class UserMediaListEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val userId: String?,
    private val username: String?
) : PagingLimitEndpoint<List<UserMediaListEntry>> {

    private var page: Int? = null
    private var limit: Int? = null

    private var category: Category? = null
    private var search: String? = null
    private var searchStart: String? = null
    private var filter: UserMediaListFilterType? = null
    private var includeHentai: Boolean? = null
    private var sort: UserMediaListSortCriteria? = null

    override fun page(page: Int?) = this.apply { this.page = page }
    override fun limit(limit: Int?) = this.apply { this.limit = limit }

    init {
        require(userId.isNullOrBlank().not() || username.isNullOrBlank().not()) {
            "You must pass either an userId or an username."
        }
    }

    /**
     * Sets the category to filter by.
     */
    fun category(category: Category?) = this.apply { this.category = category }

    /**
     * Sets the query to search for.
     */
    fun search(search: String?) = this.apply { this.search = search }

    /**
     * Sets the query to search for only from the start.
     */
    fun searchStart(searchStart: String?) = this.apply { this.searchStart = searchStart }

    /**
     * Sets the filter.
     */
    fun filter(filter: UserMediaListFilterType?) = this.apply { this.filter = filter }

    /**
     * Sets if hentai should be included in the result.
     */
    fun includeHentai(includeHentai: Boolean? = true) = this.apply { this.includeHentai = includeHentai }

    /**
     * Set the criteria for sorting.
     */
    fun sort(sort: UserMediaListSortCriteria?) = this.apply { this.sort = sort }

    override fun build(): ProxerCall<List<UserMediaListEntry>> {
        return internalApi.userMediaList(
            userId, username, category, page, limit, search, searchStart, filter, sort, includeHentai
        )
    }
}
